src/coreclr/gc/unix/numasupport.cpp

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "numasupport.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <minipal/utils.h>
+
+// The highest NUMA node available
+int g_highestNumaNode = 0;
+// Is numa available
+bool g_numaAvailable = false;
+
+#ifdef TARGET_LINUX
+static int GetNodeNum(const char* path, bool firstOnly)
+{
+	DIR *dir;
+	struct dirent *entry;
+    static int result = -1;
+
+	dir = opendir(path);
+	if (dir)
+    {
+		while ((entry = readdir(dir)) != NULL)
+        {
+			if (strncmp(entry->d_name, "node", STRING_LENGTH("node")))
+				continue;
+
+			int nodeNum = strtoul(entry->d_name + STRING_LENGTH("node"), NULL, 0);
+			if (result < nodeNum)
+				result = nodeNum;
+
+            if (firstOnly)
+                break;
+		}
+
+		closedir(dir);
+	}
+
+    return result;
+}
+#endif
+
+void NUMASupportInitialize()
+{
+#ifdef TARGET_LINUX
+	if (syscall(__NR_get_mempolicy, NULL, NULL, 0, 0, 0) < 0 && errno == ENOSYS)
+        return;
+
+    g_numaAvailable = true;
+    g_highestNumaNode = GetNodeNum("/sys/devices/system/node", false);
+#endif
+}
+
+int GetNumaNodeNumByCpu(int cpu)
+{
+#ifdef TARGET_LINUX
+    char path[64];
+    if (snprintf(path, sizeof(path), "/sys/devices/system/cpu/cpu%d", cpu) < 0)
+        return -1;
+
+    return GetNodeNum(path, true);
+#else
+    return -1;
+#endif
+}
+
+long BindMemoryPolicy(long start, unsigned long len, long nmask, int usedNodeMaskBits)
```

Comment:
This should take the actual like types for the arguments like `void* start`, so that the internal details of the syscall method signature do not leak out.